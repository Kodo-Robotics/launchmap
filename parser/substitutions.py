"""
Substitution Parser for ROS 2 Launch File AST Parser

This module extracts and formats launch-time substitution expressions used in
ROS 2 launch files, such as LaunchConfiguration, EnvironmentVariable, and
PathJoinSubstitution. It is designed to work safely using static AST parsing,
without executing any Python code.

‚úÖ Supported Substitutions (Static Parsing):
- LaunchConfiguration('name')
- LaunchConfiguration('name', default='value')
- EnvironmentVariable('VAR_NAME')
- PathJoinSubstitution([...TextSubstitution(text=...)])
- Boolean string literals ("true", "false") ‚Üí parsed as bool

‚ö†Ô∏è Partially Supported / Fallback as "<unresolved>":
- Substitutions passed as variables
- LaunchConfiguration(default=EnvironmentVariable(...)) ‚Äî no nested resolution
- PathJoinSubstitution with non-constant arguments
- Any substitution containing unknown or dynamic input

üõë Not Supported by Static Parsing:
- Command([...]) or FindExecutable(...) ‚Äî require subprocess/runtime environment
- Custom/unimported substitution types
- Nested or computed substitutions
- Default values derived from functions or variables

For unsupported or ambiguous substitutions, the parser returns symbolic
placeholders (e.g., "<unresolved>", "<unhandled:SubType>") to preserve structure
and enable graceful degradation in the visual output.

"""

import ast
from parser.context import ParseContext
from parser.utils import get_kwarg

def parse_substitution(node: ast.Call, ctx: ParseContext = None) -> str:
    if not isinstance(node, ast.Call) or not isinstance(node.func, ast.Name):
        return "<unresolved>"

    func = node.func.id

    if func == "LaunchConfiguration":
        return _parse_launch_configuration(node, ctx)
    
    elif func == "EnvironmentVariable":
        return _parse_environment_variable(node)
    
    elif func == "PathJoinSubstitution":
        return _parse_path_join(node)
    
    elif func == "TextSubstitution":
        return _parse_text_substitution(node)
    
    return f"<unhandled:{func}>"

def _parse_launch_configuration(node: ast.Call, ctx: ParseContext = None) -> str:
    name = None
    default = None

    if node.args:
        name = _resolve_const(node.args[0])
    
    for kw in node.keywords:
        if kw.arg == "name":
            name = _resolve_const(kw.value)
        elif kw.arg == "default":
            default = _resolve_const(kw.value)
    
    if name is None:
        return "<unresolved>"
    
    if ctx:
        ctx.track_launch_arg_usage(name)
        ctx.visitor.used_arguments.append(name)
    
    return f"${{LaunchConfiguration:{name}:default={default}}}" if default is not None else f"${{LaunchConfiguration:{name}}}"

def _parse_environment_variable(node: ast.Call) -> str:
    name = _resolve_const(node.args[0]) if node.args else "<unknown>"
    return f"${{EnvironmentVariable:{name}}}"

def _parse_path_join(node: ast.Call) -> str:
    if not node.args or not isinstance(node.args[0], ast.List):
        return "<unresolved>"
    
    parts = []
    for elt in node.args[0].elts:
        part = parse_literal_or_substitution(elt)
        parts.append(str(part))
    
    return f"${{PathJoinSubstitution:{'/'.join(parts)}}}"

def _parse_text_substitution(node: ast.Call) -> str:
    text = get_kwarg(node, "text")
    return _resolve_const(text) if text else "<unresolved>"

def _resolve_const(val):
    if isinstance(val, ast.Constant):
        if isinstance(val.value, str) and val.value.lower() in ("true", "false"):
            return val.value.lower() == "true"
        return val.value
    return "<unresolved>"

def parse_literal_or_substitution(node):
    if isinstance(node, ast.Constant):
        return _resolve_const(node)
    elif isinstance(node, ast.Call):
        return parse_substitution(node)
    elif isinstance(node, ast.Name):
        return "<unresolved>"
    return "<unresolved>"