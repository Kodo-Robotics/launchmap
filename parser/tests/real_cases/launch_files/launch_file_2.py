# Copyright (c) 2025 Kodo Robotics
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from launch import LaunchDescription
from launch.actions import DeclareLaunchArgument, OpaqueFunction
from launch.substitutions import LaunchConfiguration, PathJoinSubstitution
from launch_ros.actions import Node
from launch_ros.substitutions import FindPackageShare


def launch_setup(context, *args, **kwargs):
    param_path = LaunchConfiguration("driver_param_file").perform(context)
    if not param_path:
        param_path = PathJoinSubstitution(
            [
                FindPackageShare("generic_driver_pkg"),
                "params",
                "driver_config.param.yaml",
            ]
        ).perform(context)

    driver_node = Node(
        package="generic_driver_pkg",
        executable="generic_driver_node_exe",
        name="generic_driver_node",
        parameters=[param_path],
        output="screen",
        arguments=["--ros-args", "--log-level", "info", "--enable-stdout-logs"],
    )

    return [driver_node]


def generate_launch_description():
    declared_arguments = []

    def add_launch_arg(name: str, default_value: str = None, description: str = ""):
        declared_arguments.append(
            DeclareLaunchArgument(
                name, default_value=default_value, description=description
            )
        )

    add_launch_arg(
        "driver_param_file",
        default_value="",
        description="Path to the driver parameter YAML file",
    )

    return LaunchDescription(
        [*declared_arguments, OpaqueFunction(function=launch_setup)]
    )
