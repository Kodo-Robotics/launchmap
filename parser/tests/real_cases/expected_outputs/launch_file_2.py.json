{
  "arguments": [
    {
      "default_value": "",
      "description": "Path to the driver parameter YAML file",
      "name": "driver_param_file"
    }
  ],
  "opaque_functions": [
    {
      "name": "launch_setup(context, *args, **kwargs)",
      "returns": {
        "nodes": [
          {
            "package": "generic_driver_pkg",
            "executable": "generic_driver_node_exe",
            "name": "generic_driver_node",
            "parameters": [
              "${LaunchConfiguration:driver_param_file}"
            ],
            "output": "screen",
            "arguments": [
              "--ros-args",
              "--log-level",
              "info",
              "--enable-stdout-logs"
            ]
          }
        ]
      }
    }
  ],
  "launch_argument_usages": [
    {
      "argument": "driver_param_file",
      "path": "opaque_functions[0].returns.nodes[0].parameters[0]"
    }
  ]
}